fix(#136): memory pipeline hardening + issue relinking (#135) (#138)
medic/cht-agent: docs/memory-pipeline-decisions.md

# Memory pipeline: design decisions

> **Status:** DRAFT — awaiting sign-off (posted to PR #138 on 2026-08-27)

Reviewer: **suggestion (non-blocking):** Merging this PR lands a document that says nothing in it is
decided. But the D3 and D6 code in the same PR already treats it as the spec. This review's
body records one decider's reactions to all of D1-D11. Point the status line at the recorded
reactions, and keep it honest about who has reacted:

```suggestion
**Status:** DRAFT — decider reactions are recorded in the #138 reviews: the round-3 review approves D1–D11 with amendments noted there; reactions from the other deciders are pending. An approving review on #138 also accepts the D3 and D6 implementations this PR ships.
```

**Deciders:** @alexosugo, @sugat009, @Hareet
**Context:** PR #138 review (rounds 1–3). The review surfaced questions that are
product decisions, not code defects. This doc records one decision per question so
future reviews argue against a written spec instead of re-negotiating in threads.
Each decision below is a **proposal** until a decider objects or approves.

A fact that frames everything: the nightly pipeline has never produced a draft in

> production. All 50 scheduled runs since 2026-06-24 either found no PRs or died on

Reviewer: **nitpick (non-blocking):** The count is wrong. 64 scheduled runs existed on this doc's date, 68
today. The verifiable core stands: zero drafts have ever been committed.

```suggestion
production. Every scheduled run since 2026-06-24 (64 by this doc's date) either found no PRs or died on
```

- `[skip ci]` on the pipeline commit is removed; the `check-pending` guard runs
  on the draft PR — closing the "guard never runs on the commit that adds a
  draft" gap (Sugat's note on `unit_tests.yml`).
- The nightly needs `permissions: pull-requests: write` and a branch-naming
  convention (`memory/nightly-<date>`), plus dedup of an already-open nightly PR
  (append to the existing branch rather than opening a second PR).

## D2. CI gets `OPENROUTER_API_KEY`; the workflow also wires `ANTHROPIC_API_KEY`

**Decision:** Add the repo secret and pass both env vars in `run-pipeline.yml`,
since `distiller.ts` supports both and the error message advertises both.
Owner: whoever holds repo admin (secret cannot be added from a PR).

## D3. Base-branch gate: exclusion is a skip, never an error

*(Blocker on `scraper.ts`; fix agreed in review, recorded here as the contract)*

**Decision:** A PR merged into a non-default branch is a normal, auditable
exclusion. It must write a `_skipped.ndjson` row and leave the run's exit code
untouched. Implementation: carry `baseRefName` on `ScrapedPR` and decide in the
filter (`checkSkipRules`), not by throwing from the scraper. `PR #N is not

> merged` moves to the same channel. A pipeline-level test asserts

Reviewer: **issue (blocking):** This sentence is false at this PR's own head. `fetchAndParseMetadata`
still throws `ScraperError` for an unmerged PR. The throw is correct. An unmerged PR cannot
become a `ScrapedPR`, and only an explicit `--pr` reaches it. Record the deviation instead of
promising the move:

```suggestion
filter (`checkSkipRules`), not by throwing from the scraper. `PR #N is not
merged` stays a thrown error: an unmerged PR cannot become a `ScrapedPR`, and only
an explicit `--pr` reaches it, where a loud operator error is the right signal.
Note `--force` bypasses the base-branch gate, like every filter rule. A pipeline-level test asserts
```

`failures === 0` plus one appended skip row for a batch containing an excluded PR.

**Contract:** an exclusion the design intends is never allowed to fail the run
or to skip the audit log. Anything that increments `failures` must be an actual
malfunction (gh failure, corrupt JSON, rate limit).

## D4. Release branches (`5.1.x` style) are IN scope; the gate allows them

*(Sugat Q1)*

**Decision:** The gate accepts the default branch plus branches matching
`/^\d+\.\d+\.x$/`. Everything else is skipped (per D3).

**Rationale:** cht-core PR 11137 merged into `5.1.x`, shipped in tags 5.1.3+,
and never reached `master`. That is real released work; excluding it loses
knowledge the corpus exists to hold. The backport-duplication risk is already
handled: dedup keys on issue id and keeps the mainline fix, so a `master` fix

> plus its release-branch backports collapse to one entry.

Reviewer: **issue (non-blocking):** The example does not show lost knowledge. PR 11137 (`5.1.x`, issue
#11128) has a master twin, #11133, merged the same day. The decision may still be right. The
rationale needs an honest example or none. Also, "keeps the mainline fix" describes D7's
deferred interim tiebreak, not the shipped code. `dedupeByIssueId` ranks strong `domainFit`
first, then lowest PR number.

```suggestion
**Rationale:** release branches carry shipped work, and the gate widening is cheap.
The backport-duplication risk is already handled: dedup keys on issue id, so a
`master` fix plus its release-branch backports collapse to one entry. Today the
tiebreak is strong `domainFit`, then lowest PR number; the mainline-first tiebreak
is D7's interim rule and lands with D7. (PR 11137 is not an example of lost
knowledge: its master twin #11133 fixes the same issue #11128 and merged the same day.)
```

`BatchState` initialization).

## D10. `_skipped.ndjson` truncation in #138 was deliberate

*(Sugat Q4)*

**Decision:** Confirmed. All 7 removed rows were `prNumber: 1` test pollution
("LLM triage skipped") written by a spec that ran without a `logPath` (#146).
The remainder of #146 (pass `logPath` in the four option-less `filterPR` spec
calls, `NODE_ENV=test` guard) stays on #146 and does not grow #138.

**Post-rebase note (2026-08-27):** #158 landed on `main` and gitignores
`agent-memory/_skipped.ndjson`. After the rebase the file is no longer in the #138
diff at all, so the truncation question is closed by `main`, not by this PR.

## D11. Scope discipline for #138 and after

**Decision:** #138 takes only the D3 and D6 fixes (its two in-PR blockers) plus
the D7 interim tiebreak if reviewers want it in the same PR; otherwise D7 ships
with #154. Everything else here lands as follow-up PRs referencing this doc:
D1+D2 (pipeline-to-PR), D4 (release-branch gate), D5 (no change needed — it is
the current behaviour, kept), D8 (CLI guard). Non-blocking review findings
(reconcile buckets, warn surfacing via `linkage_warning:` frontmatter, transient

> `defaultBranch` cache, stale runbook lines) become small issues, not #138 commits.

Reviewer: **suggestion (non-blocking):** Three review findings have no written disposition, against this
doc's own one-decision-per-question rule: promote-commit staging of `_pending` deletions, dry-run
visibility of dedup drops, and the `DRAFT_SHAPE` tightening. Add them:

```suggestion
(reconcile buckets, warn surfacing via `linkage_warning:` frontmatter, transient
`defaultBranch` cache, stale runbook lines, promote-commit staging of `_pending`
deletions, dry-run visibility of dedup drops, and the `DRAFT_SHAPE` tightening)
become small issues, not #138 commits.
```

  "processes every PR and does not exit when all succeed" and "records a failure and
  exits 1 when a PR throws". The D3 test is a third sibling. No new harness.
- **It ends "which channel" arguments.** The open blockers are all about channel:
  error versus skip, delete versus move, log versus surface. A boundary test names the
  channel in its assertion.
- **It gives the review a citation.** A reviewer writes "covered by
  `runPipeline › skips a non-default-base PR without failing the batch`" instead of
  re-tracing the code.
- **It exposes tests that cannot fail.** Delete the guard, run the test, watch it
  fail. That mutation check is how the `scraper.spec.ts:178` gap was found.

### Applied to the two open blockers

**D3**

```
runPipeline([excludedPR, normalPR])
  → failures === 0
  → process.exit not called
  → _skipped.ndjson gained one row: { prNumber: excludedPR, decision: 'skip', reason: /non-default branch/ }
  → normalPR still produced a draft
```

**D6**

```
openReviewPR(domain with two same-issue drafts)
  → canonical promoted
  → dropped draft exists at _pending/_collapsed/<domain>/<file>
  → dropped draft no longer in _pending/<domain>/

>   → audit row: { decision: 'skip', reason: /duplicate of .* "<dropped title>"/ }

Reviewer: **issue (blocking):** The code writes `decision: 'flag-for-human'` (`writeSkipEntry`,
`open-review-pr.ts:161`) and the shipped test asserts it. The sketch says `skip`.
`flag-for-human` is the better channel: a human may merge `_collapsed/` content by hand. Fix the
sketch:

```suggestion
  → audit row: { decision: 'flag-for-human', reason: /duplicate of .*; ".*" moved to _collapsed/ }
```
